### PR TITLE
Addressing issue #154

### PR DESCRIPTION
**Pull request by flixman**
As the sending of the ZFS streams is per stream, we can disable children ZFS from being snapshot/sent by setting the property "org.znapzend:enabled" to "off" on them. For the dataset being processed, in case the recursive flag is set, its children will be checked and their recently created snapshots removed.

### CODE REVIEW

File: lib/ZnapZend.pm [hunk 1]
```diff
@@ -427,6 +427,42 @@ my $createSnapshot = sub {
...
+                
+                # get the value for org.znapzend property
+                my @cmd = (@{$self->zZfs->priv}, qw(zfs get -H -s local -o value org.znapzend:enabled), $dataSet);
+                print STDERR '# ' . join(' ', @cmd) . "\n" if $self->debug;
+                open my $prop, '-|', @cmd;
+
+                # if the property does not exist, the command will just return. In this case,
+                # the value is implicit "on"
+                $prop = <$prop> || "on";
+                if ( substr($prop, 0, length('off')) eq 'off' ) {
```
Comment (oetiker) [inline lib/ZnapZend.pm]: `if ( $prop eq 'off')` should be fine :)

File: lib/ZnapZend.pm [hunk 2]
```diff
@@ -427,6 +427,42 @@ my $createSnapshot = sub {
...
 
+    # remove snapshots from descendant subsystems that have the property "enabled" to "off", if the
+    # "recursive" flag is set to "on"
+    if ($backupSet->{recursive} eq 'on') {
+
+        $self->zLog->info("checking ZFS dependent datasets from '$backupSet->{src}' explicitely excluded");
+
+        # restrict the list to the datasets that are descendant from the current
+        my @dataSetList = grep /^$backupSet->{src}($|\/)/, @{$self->zZfs->listDataSets()};
+        if ( scalar @dataSetList > 0 ) {
```
Comment (oetiker) [inline lib/ZnapZend.pm]: `if (@dataSetList)` is fine

File: lib/ZnapZend.pm [hunk 3]
```diff
@@ -427,6 +427,42 @@ my $createSnapshot = sub {
...
+                # if the property does not exist, the command will just return. In this case,
+                # the value is implicit "on"
+                $prop = <$prop> || "on";
+                if ( substr($prop, 0, length('off')) eq 'off' ) {
+                    push(@dataSetsExplicitelyDisabled, $dataSet . '@' . $snapshotSuffix);
+                }
+            }
+
+            # remove the snapshots previously marked
+           if ( scalar @dataSetsExplicitelyDisabled > 0 ){
```
Comment (oetiker) [inline lib/ZnapZend.pm]: `if (@dataSetsExplicielyDisabled)`

File: lib/ZnapZend.pm (hunk 1 above)
Comment (flixman) [reply lib/ZnapZend.pm]: I have added a chomp also, because the command returns with a '\n'.

**Comment by flixman** [comment]
Seems the project has a build error with perl 5.26, but is an amazon s3 error :-S @oetiker ; I think I can do nothing to get this fixed?

**Comment by oetiker** [comment]
Sorry I am on the road, but I get the impression that the error is from a missing module: Test::Exception ... adding this to PERL_MODULES could fix the problem

**Comment by flixman** [comment]
But, AFAIK, I can not modify travis configuration, right? (Just to confirm whether I can or I cannot fix it)

**Comment by oetiker** [comment]
you can modify everything ... just edit .travis.yml ... this is what gets used to test your PR

**Comment by oetiker** [comment]
thanks

**Comment by oetiker** [comment]
refers to #154

**Pull request merged**